feat(AIP-4235): automatically populate fields in the request message (#1307)
googleapis/aip: aip/client-libraries/4235.md

---
id: 4235
state: draft
created: 2024-02-15
---

# Automatically populate fields in the request message

> For APIs that leverage request idempotency as described via [AIP-155], it could be useful to have the client libraries autogenerate the `request_id` if it is not already set by the customer.

Reviewer: I know this is kind of annoying, but please wrap lines to 80 columns :D

Author: Ahh good call - updated!

**Note:** This feature is primarily written for `request_id` fields within the request message. However, this feature should work for a field of any name, as long as the below conditions hold true.

## Guidance

APIs **may** configure fields in the request message for automatic population.

> If an API chooses to do so, they **must** guarantee idempotency for the configured fields.

Reviewer: What do you mean by "guarantee idempotency"? How does a service do that for a request field? If you are referring to AIP-155, let's not get into server behavior here.

Author: Yes - I was getting tripped up over how much this AIP should provide the same guidance as AIP-155. I'll remove this and the below section.

- If a duplicate request is detected, the server **should** return the
  response for the previously successful request, because the client most
  likely did not receive the previous response.

> - APIs **may** choose any reasonable timeframe for honoring request IDs.

Reviewer: Let's not get into the AIP-155 guidance here - splitting guidance between two AIPs is asking for them to be out of sync.

For a field to be automatically populated, the below configurations **must** be true:

> - The field **must** exist as a top-level string field of a unary method's request message (e.g. nested fields will not be eligible for automatic population).

Reviewer: Can you separate this into three different requirements? e.g.,

```
- The field **must** be of type `string`
- The field **must** be at the top level of the request i.e., nested fields are not supported
- The RPC **must** be a unary RPC i.e., streaming RPCs are not supported
```

This way we can implement three different linter rules and guidance is clearly separated out.

> - The unary method which has the field in the request message **must** be explicitly designated in [google.api.client.Publishing.method_settings.selector] using the fully qualified name of the method.

Reviewer: I don't think this guidance is necessary. Just saying it needs to be in the MethodSettings alludes to the fact that that method needs to have settings in the config.

> - The field **must** *not* be annotated with [google.api.field_behavior = REQUIRED].

Reviewer: ```suggestion
- The field **must** *not* be annotated with [`google.api.field_behavior = REQUIRED`][required].
```

> - The field name **must** be listed in [google.api.client.Publishing.method_settings.auto_populated_fields].

Reviewer: I'd put this last since all other guidance is proto based

> - The field name **must** be listed in [google.api.client.Publishing.method_settings.auto_populated_fields].

Reviewer: ```suggestion
- The field name **must** be listed in the [`google.api.MethodSettings.auto_populated_fields`][apf] entry of the service config for the target method.
```

Author: Quick q  on this - how were you able determine the naming to be `google.api.MethodSettings.auto_populated_fields`  and skip the `client` and `Publishing` pieces altogether?

Author: Chatted offline and refactored to:
`The field name **must** be listed in the [`google.api.MethodSettings.auto_populated_fields`][apf] of an entry
  in [`google.api.Publishing.method_settings`][apf] for the target method.`

> - The field **must** be annotated with [google.api.field_info.format = UUID4].

Reviewer: ```suggestion
- The field **must** be annotated with [`(google.api.field_info).format = UUID4`][uuid4].
```

**Note:** This feature currently only supports automatic population of string UUID4 values. It may be expanded in the future to support other format types.

> ### Generator configuration

Reviewer: This isn't really generator configuration, the previous section was. This is more generator & client library _behavior_

Author: Updated the heading

> Client library generators **should** enable automatic population of request fields at the time of sending the request in the specified format if the above configurations are true.
> If a field is not configured as above, *or* it is set in the request by the customer,
> then the field **must not** be automatically populated.

Reviewer: ```suggestion
If the aforementioned requirements are met for a given field, client library generators **should**
enable automatic population said field in the generated client. If a field is specified in the
`auto_populated_fields`, but does not meet the structural requirements, the client library generators
**must not** enable automatic population for that field. Client library generators **may** emit an error
during generation.
```

> **If** a client library generator supports automatic population of request fields, then those automatically populated fields **must** be reused for retries of the same request.

Reviewer: ```suggestion
Client libraries **must** reuse auto-populated values for retries of the same request i.e., the auto-populated
fields **must not** be regenerated for each RPC attempt with a single request message.
```

As of February 15, 2024, this feature is supported in the Node and Java client library generators. 

[AIP-155]: https://google.aip.dev/155

> [google.api.client.Publishing.method_settings.selector]: https://github.com/googleapis/googleapis/blob/master/google/api/client.proto

Reviewer: As with the guidance above, remove this as it won't be needed.

> [google.api.client.Publishing.method_settings.auto_populated_fields]: https://github.com/googleapis/googleapis/blob/master/google/api/client.proto

Reviewer: ```suggestion
[apf]: https://github.com/googleapis/googleapis/blob/master/google/api/client.proto
```

> [google.api.field_info.format = UUID4]: https://github.com/googleapis/googleapis/blob/master/google/api/field_info.proto

Reviewer: ```suggestion
[uuid4]: https://google.aip.dev/202#uuid4
```

> [google.api.field_behavior = REQUIRED]: https://github.com/googleapis/googleapis/blob/master/google/api/field_behavior.proto

Reviewer: ```suggestion
[required]: https://google.aip.dev/203#required
```

---
id: 4235

> state: draft

Reviewer: ```suggestion
state: approved
```

> created: 2024-02-15

Reviewer 2: ```suggestion
created: 2024-03-07
```

---

# Automatically populate fields in the request message

For APIs that leverage request idempotency as described via [AIP-155], APIs may
choose to have the client libraries automatically populate the `request_id`
if it is not already set by the customer.

**Note:** This feature is primarily written for `request_id` fields within the
request message. Nonetheless, this feature **must** work for a field of any
name, as

> long as the below conditions hold true.

Reviewer: ```suggestion
name, as long as the below conditions hold true.
```
fix line wrapping

## Guidance

APIs **may** configure fields in the request message for automatic population.
For a field to be automatically populated, **all** the below configurations
**must** be true:

- The field **must** be of type `string`
- The field **must** be at the top-level of the request message
- The RPC **must** be a unary RPC (i.e. streaming RPCs are not supported)
- The field **must not** be annotated
  with [`google.api.field_behavior = REQUIRED`][required].
- The field **must** be annotated
  with [`google.api.field_info.format = UUID4`][uuid4].
- The field name **must** be listed
  in the [`google.api.MethodSettings.auto_populated_fields`][apf] entry
  in [`google.api.Publishing.method_settings`][apf]
  for the target method.

### Expected Generator and Client Library Behavior

If the aforementioned requirements are met for a given field, client library
generators **must**

> enable automatic population of said field in the generated client.

Reviewer: ```suggestion
If the aforementioned requirements are met for a given field, client library
generators **must** enable automatic population of said field in the
generated client.
```

fix line wrapping

If a field is specified in the `auto_populated_fields`, but does not meet the
structural requirements, the client library generators **must not** enable
automatic population for that field. Client library generators **may** emit an
error during generation.

Client libraries **must** reuse automatically populated values for retries of
the same request i.e., the automatically populated
fields **must not** be regenerated for each RPC attempt with a single request

> message.

Reviewer: ```suggestion
Client libraries **must** reuse automatically populated values for retries of
the same request i.e., the automatically populated fields **must not** be
regenerated for each RPC attempt with a single request message.
```

fix line wrapping

---
id: 4235
state: draft
created: 2024-02-15
---

# Automatically populate fields in the request message

For APIs that leverage request idempotency as described via [AIP-155], it could
be useful to have the client libraries automatically populate the `request_id`
if it is
not already set by the customer.

**Note:** This feature is primarily written for `request_id` fields within the
request message. However, this feature should work for a field of any name, as
long as the below conditions hold true.

## Guidance

APIs **may** configure fields in the request message for automatic population.

For a field to be automatically populated, the below configurations **must** be
true:

- The field **must** be of type `string`
- The field **must** be at the top-level of the request message (e.g. nested

>   fields are not supported)

Reviewer 3: I feel it may not be necessary to mention the cases not covered by the AIP, but if that is a common practice for AIPs, please disregard this comment.

Author: That's a fair point - I added the comment more to help clarify what "top-level of request message" means, rather than to mention it as a case not covered by the AIP. 
But I checked a couple other AIPs that use the term `top-level`, and it seems commonly used enough that I don't think I need this extra clarification, so I've removed it.

---
id: 4235
state: draft
created: 2024-02-15
---

# Automatically populate fields in the request message

For APIs that leverage request idempotency as described via [AIP-155], it could
be useful to have the client libraries automatically populate the `request_id`
if it is

> not already set by the customer.

Reviewer 2: ```suggestion
For APIs that leverage request idempotency as described via [AIP-155], client libraries must automatically populate the `request_id` if it is not already set by the customer.
```

Author: Same question as above - since it's not implemented by all languages (yet), I'm not sure we can have this wording.

Author: @parthea  So I refactored it slightly to read "APIs may choose to have the client libraries automatically populate the `request_id` if it is not already set by the customer." as I think the purpose of this AIP is to provide services a way to configure the client library to automatically populate the `request_id`, but that it won't automatically get automatically populated if they *only* follow AIP-155. 

Let me know if you think this is insufficient.

Reviewer 2: Looks good

**Note:** This feature is primarily written for `request_id` fields within the

> request message. However, this feature should work for a field of any name, as

Reviewer 3: Maybe we should avoid using the word "should" in AIPs outside of its technical sense: https://google.aip.dev/8#requirement-keywords

Reviewer: ```suggestion
request message. However, this feature could work for a field of any name, as
```

+1 here is a suggestion

Reviewer 2: How about `this feature must work for a field of any name, as...` instead if we want to ensure that the feature works generally for all fields

Reviewer: sgtm

> long as the below conditions hold true.

Reviewer 2: Is this note needed? Right now we only have guidance for the `request_id` field

Reviewer: I think the aspirational side of me wants to say that we should keep this, because there is no reason why, if all of the conditions are met, we can't auto populate a field named something other than `request_id`. I would hope that most implementations aren't looking specifically for a `request_id` field by name, given that technically any field name could appear in the config - this can definitely be generalized software feature.

Author: Kept the note and changed the wording slightly based on the below comment

## Guidance

APIs **may** configure fields in the request message for automatic population.

> For a field to be automatically populated, the below configurations **must** be

Reviewer 3: maybe emphasize **all** of the configurations (or conditions?) must be true.

true:

- The field **must** be of type `string`
- The field **must** be at the top-level of the request message (e.g. nested
  fields are not supported)
- The RPC **must** be a unary RPC (i.e. streaming RPCs are not supported)
- The field **must not** be annotated
  with [`google.api.field_behavior = REQUIRED`][required].
- The field **must** be annotated
  with [`google.api.field_info.format = UUID4`][uuid4].
- The field name **must** be listed

>   in the [`google.api.MethodSettings.auto_populated_fields`][apf] of an entry

Reviewer 2: ```suggestion
  in the [`google.api.MethodSettings.auto_populated_fields`][apf] entry
```

  in [`google.api.Publishing.method_settings`][apf]
  for the target method.

**Note:** This feature currently only supports automatic population of string

> UUID4 values. It may be expanded in the future to support other format types.

Reviewer 2: Is this note needed if we have a requirement that the field 
>**must** be of type `string`

Author: Good point. I was thinking that we might want to indicate we're open to other formats if there's a need, but I think better to just explain what's currently approved/implemented. I've removed the note.

### Expected Generator and Client Library Behavior

If the aforementioned requirements are met for a given field, client library

> generators **should**

Reviewer 2: ```suggestion
generators **must**
```

Author: So I wasn't sure about this one since I think it's not technically implemented in all language generators yet. @noahdietz do you know if it is okay to update this wording to **must** now, or should we leave it as is until we've confirmed it's implemented for all languages?

Reviewer: I think we can change it to `**must**`. We at least have it planned everywhere, so let's "code" to the ideal state of "must".

Author: Cool, done
